vhost-device-input/src/vhu_input.rs

```diff
@@ -0,0 +1,625 @@
+    }
+}
+
+pub(crate) struct VuInputBackend<T: InputDevice> {
+    event_idx: bool,
+    ev_dev: T,
+    pub exit_event: EventFd,
+    select: u8,
+    subsel: u8,
+    mem: Option<GuestMemoryAtomic<GuestMemoryMmap>>,
+    ev_list: VecDeque<VuInputEvent>,
+}
+
+impl<T: InputDevice> VuInputBackend<T> {
+    pub fn new(ev_dev: T) -> std::result::Result<Self, std::io::Error> {
+        Ok(VuInputBackend {
+            event_idx: false,
+            ev_dev,
+            exit_event: EventFd::new(EFD_NONBLOCK).map_err(|_| VuInputError::EventFdError)?,
+            select: 0,
+            subsel: 0,
+            mem: None,
+            ev_list: VecDeque::new(),
+        })
+    }
+
+    fn process_event(&mut self, vring: &VringRwLock) -> Result<bool> {
+        let mut index = 0;
+        let mut last_sync_index = 0;
+
+        for event in self.ev_list.iter() {
+            index += 1;
+            if event.ev_type == EV_SYN as u16 && event.code == SYN_REPORT as u16 {
+                last_sync_index = index;
+            }
+        }
+
+        index = 0;
+        while index < last_sync_index {
+            let descriptors: Vec<_>;
+            let descriptor;
+            let event = self.ev_list.get(index).unwrap();
+            index += 1;
+
+            let mem = self.mem.as_ref().unwrap().memory();
+
+            let desc = vring
+                .get_mut()
+                .get_queue_mut()
+                .pop_descriptor_chain(mem.clone());
+
+            if let Some(desc_chain) = desc {
+                descriptors = desc_chain.clone().collect();
+                if descriptors.len() != 1 {
+                    return Err(VuInputError::UnexpectedDescriptorCount(descriptors.len()));
+                }
+                descriptor = descriptors[0];
+
+                let len = desc_chain
+                    .memory()
+                    .write(event.to_bytes().as_slice(), descriptor.addr())
+                    .unwrap();
+
+                if vring.add_used(desc_chain.head_index(), len as u32).is_err() {
+                    println!("Couldn't write event data to the ring");
+                    return Err(VuInputError::UnexpectedWriteVringError);
+                }
+            } else {
+                // Now cannot get avaliable descriptor, which means the host cannot process
+                // event data in time and overrun happens in the backend. In this case,
+                // we simply drop the incomping input event and notify guest for handling
+                // events. At the end, it returns Ok(false) so can avoid exiting the thread loop.
+                self.ev_list.clear();
+
+                vring
+                    .signal_used_queue()
+                    .map_err(|_| VuInputError::SendNotificationFailed)?;
+
+                return Ok(false);
+            }
+        }
+
+        index = 0;
+        while index < last_sync_index {
+            self.ev_list.remove(0);
+            index += 1;
+        }
```

Comment:
```suggestion
        for i in 0..last_sync_index {
            self.ev_list.remove(i);
        }
```

Reply:
Or, `self.ev_list.drain(0..last_sync_index)`?

Reply:
Yeah, I used @Ablu 's method for cleanup the event list.  As I said above, here `last_sync_index` is an inclusive value, so we should use `last_sync_index + 1` for the range.